In [20]:
import pandas as pd
df_orders = pd.read_csv('orders_with_unitPrice.csv')
df_cons = pd.read_csv('cleaned_consumption.csv')

In [11]:
df_orders

Unnamed: 0  order_id  product_category  product_id  order_date  \
0              0         0                 4          61  2020-01-02   
1              1         1                 4          58  2020-01-04   
2              2         2                 0          12  2020-02-04   
3              3         3                 4          46  2020-03-17   
4              4         4                 4          50  2020-04-16   
...          ...       ...               ...         ...         ...   
1802        1802      1837                 4          50  2025-01-17   
1803        1803      1838                 4          48  2025-01-17   
1804        1804      1839                 4          48  2025-01-17   
1805        1805      1889                 2         106  2025-04-03   
1806        1806      1893                 4          32  2025-05-29   

     delivery_date        amount   qty  valid_delivered_qty     unit_price  
0       2020-01-14  1.249730e+05   1.0                  1.0  124972.990282  
1       2020-01-14  1.202992e+06   4.0                  4.0  300748.009134  
2       2020-02-19  3.010961e+06   5.0                  5.0  602192.263763  
3       2020-03-29  5.144759e+06  25.0                 12.0  205790.366021  
4       2020-04-16  6.565086e-01   1.0                  1.0       0.656509  
...            ...           ...   ...                  ...            ...  
1802           NaN  4.060958e+05   2.0                  NaN  203047.913058  
1803           NaN  1.094812e+06  10.0                  NaN  109481.239608  
1804           NaN  2.189625e+05   2.0                  NaN  109481.239608  
1805           NaN  1.121301e+06   6.0                  NaN  186883.441232  
1806           NaN  1.674402e+07  50.0                  NaN  334880.375079  

[1807 rows x 10 columns]

Для каждой категории выявить процент не дошедших товаров 

In [16]:
df_orders.info()

<class 'pandas.DataFrame'>
RangeIndex: 1807 entries, 0 to 1806
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Unnamed: 0           1807 non-null   int64  
 1   order_id             1807 non-null   int64  
 2   product_category     1807 non-null   int64  
 3   product_id           1807 non-null   int64  
 4   order_date           1807 non-null   str    
 5   delivery_date        1669 non-null   str    
 6   amount               1807 non-null   float64
 7   qty                  1807 non-null   float64
 8   valid_delivered_qty  1669 non-null   float64
 9   unit_price           1807 non-null   float64
dtypes: float64(4), int64(4), str(2)
memory usage: 141.3 KB


In [48]:
df_orders['amount'].count()

np.int64(1807)

In [67]:
for categ, group in df_orders.groupby("product_category"):
    group = group.dropna(subset=["valid_delivered_qty"])
    qty = group["qty"]
    qty_sum = qty.sum()
    no_valid_qty = (qty - group["valid_delivered_qty"])
    no_valid_qty_sum = (no_valid_qty > 0).sum()
    print(
        f"category {categ}: \nколичество недошедшего товара:",
        f"{no_valid_qty.sum()}({(no_valid_qty.sum() * 100) / qty_sum:.2f}%)",
        f"\nколичество не дошедших заказов {no_valid_qty_sum}({no_valid_qty_sum * 100 / qty.count():.2f}%)",
    )


category 0: 
количество недошедшего товара: 76.0(3.21%) 
количество не дошедших заказов 9(11.84%)
category 1: 
количество недошедшего товара: 0.0(0.00%) 
количество не дошедших заказов 0(0.00%)
category 2: 
количество недошедшего товара: 443.0(2.95%) 
количество не дошедших заказов 32(5.41%)
category 3: 
количество недошедшего товара: 3580.0(0.80%) 
количество не дошедших заказов 4(5.19%)
category 4: 
количество недошедшего товара: 752.0(4.35%) 
количество не дошедших заказов 49(5.37%)


Давайте посмотрим детальнее на категорию 0, 2 и 4

In [83]:
group = df_orders[df_orders['product_category'] == 0]
qty = group["qty"]
mask = ((qty - group["valid_delivered_qty"]) > 0)
group[mask]

Unnamed: 0  order_id  product_category  product_id  order_date  \
534          534       534                 0          16  2022-07-05   
716          716       716                 0          11  2023-04-06   
792          792       792                 0          14  2023-07-19   
821          821       821                 0          13  2023-08-22   
1097        1097      1097                 0          17  2024-04-10   
1099        1099      1099                 0          18  2024-04-10   
1376        1376      1376                 0          29  2024-11-20   
1469        1469      1469                 0          14  2025-01-31   
1600        1600      1600                 0          27  2025-04-01   

     delivery_date        amount   qty  valid_delivered_qty    unit_price  
534     2022-09-28  4.373822e+07  35.0                 34.0  1.249663e+06  
716     2023-07-06  2.101674e+07  23.0                 22.0  9.137713e+05  
792     2023-10-24  6.514194e+07  44.0                  1.0  1.480499e+06  
821     2024-01-26  2.772175e+07  14.0                 13.0  1.980125e+06  
1097    2024-09-20  3.835612e+07  19.0                 17.0  2.018743e+06  
1099    2024-09-16  5.854355e+07  29.0                 27.0  2.018743e+06  
1376    2025-05-02  2.168499e+07  68.0                 65.0  3.188969e+05  
1469    2025-04-15  1.375180e+08  46.0                 32.0  2.989521e+06  
1600    2025-06-24  6.422097e+07  92.0                 83.0  6.980541e+05

Если в категории 2 убрать заказ order_id=792, то категория по количетству досталенных товаров становиться крайне стабильной.

In [82]:
group = df_orders[df_orders['product_category'] == 2]
qty = group["qty"]
mask = ((qty - group["valid_delivered_qty"]) > 0)
group[mask]

Unnamed: 0  order_id  product_category  product_id  order_date  \
559          559       559                 2          75  2022-08-05   
761          761       761                 2         117  2023-06-16   
765          765       765                 2          85  2023-06-16   
768          768       768                 2         124  2023-06-16   
846          846       846                 2          81  2023-09-13   
1083        1083      1083                 2         109  2024-04-08   
1150        1150      1150                 2         124  2024-06-04   
1151        1151      1151                 2         125  2024-06-04   
1172        1172      1172                 2          93  2024-08-02   
1188        1188      1188                 2         109  2024-08-27   
1193        1193      1193                 2          93  2024-09-03   
1194        1194      1194                 2          94  2024-09-03   
1399        1399      1399                 2         109  2024-12-06   
1422        1422      1422                 2          93  2024-12-12   
1471        1471      1471                 2          80  2025-02-02   
1472        1472      1472                 2          81  2025-02-02   
1474        1474      1474                 2          85  2025-02-02   
1475        1475      1475                 2          86  2025-02-02   
1477        1477      1477                 2         101  2025-02-02   
1478        1478      1478                 2         102  2025-02-02   
1481        1481      1481                 2         117  2025-02-02   
1482        1482      1482                 2         118  2025-02-02   
1484        1484      1484                 2         124  2025-02-02   
1485        1485      1485                 2         125  2025-02-02   
1513        1513      1513                 2         101  2025-02-26   
1526        1526      1526                 2          83  2025-03-04   
1564        1564      1564                 2          85  2025-03-11   
1565        1565      1565                 2          86  2025-03-11   
1586        1586      1586                 2         101  2025-03-24   
1642        1642      1642                 2          83  2025-04-16   
1664        1664      1664                 2         125  2025-05-16   
1689        1689      1689                 2          87  2025-06-06   

     delivery_date        amount    qty  valid_delivered_qty     unit_price  
559     2022-09-14  5.133862e+06   50.0                 49.0  102677.239514  
761     2023-09-25  6.323675e+06   97.0                 76.0   65192.525702  
765     2023-09-24  5.273557e+06   71.0                 61.0   74275.454517  
768     2023-09-24  3.712281e+06   70.0                 62.0   53032.584372  
846     2024-01-18  3.991294e+06   26.0                 13.0  153511.323066  
1083    2024-06-03  2.748699e+06   28.0                 27.0   98167.831173  
1150    2024-07-18  1.584197e+06   25.0                 12.0   63367.884997  
1151    2024-07-18  3.168394e+05    5.0                  4.0   63367.884997  
1172    2024-10-31  2.373470e+06   15.0                 14.0  158231.348790  
1188    2025-04-17  2.560788e+06   20.0                  2.0  128039.399873  
1193    2024-10-06  4.889463e+06   34.0                 32.0  143807.734776  
1194    2024-10-06  3.163770e+06   22.0                 18.0  143807.734776  
1399    2025-06-15  1.156155e+07   68.0                 67.0  170022.767734  
1422    2025-04-06  3.133813e+07  183.0                182.0  171246.605867  
1471    2025-03-30  3.067114e+06    9.0                  2.0  340790.493948  
1472    2025-03-30  1.329252e+07   56.0                 55.0  237366.513197  
1474    2025-03-30  1.851459e+06   14.0                  9.0  132247.057353  
1475    2025-03-30  4.204778e+05    4.0                  2.0  105119.455845  
1477    2025-03-30  5.571331e+06   53.0                 46.0  105119.455845  
1478    2025-03-30  2.999295e+06   29.0                 27.0  103423.

**Категория 2:** Если в 2022-2023 годах недовозы были мелкими, то к 2025 году они стали критическими. Это говорит о нарастающем дефиците запчастей этой категории.

In [81]:
group = df_orders[df_orders['product_category'] == 4]
qty = group["qty"]
mask = ((qty - group["valid_delivered_qty"]) > 0)
group[mask]

Unnamed: 0  order_id  product_category  product_id  order_date  \
3              3         3                 4          46  2020-03-17   
33            33        33                 4          21  2020-08-11   
378          378       378                 4          61  2021-11-01   
432          432       432                 4          26  2022-03-02   
433          433       433                 4           1  2022-03-02   
434          434       434                 4          71  2022-03-02   
436          436       436                 4          61  2022-03-02   
438          438       438                 4          32  2022-03-02   
441          441       441                 4          22  2022-03-02   
444          444       444                 4          26  2022-03-02   
449          449       449                 4          45  2022-03-02   
450          450       450                 4          58  2022-03-02   
459          459       459                 4          37  2022-04-19   
482          482       482                 4          21  2022-03-02   
521          521       521                 4           0  2022-06-28   
526          526       526                 4          70  2022-06-28   
601          601       601                 4          58  2022-10-19   
618          618       618                 4          54  2022-11-09   
619          619       619                 4          56  2022-11-09   
625          625       625                 4          57  2022-11-14   
656          656       656                 4          56  2022-07-11   
671          671       671                 4          58  2023-01-25   
673          673       673                 4          32  2022-11-03   
699          699       699                 4          58  2023-02-14   
731          731       731                 4          61  2023-04-18   
770          770       770                 4          70  2023-06-21   
783          783       783                 4          71  2023-07-14   
830          830       830                 4          54  2023-08-31   
831          831       831                 4          58  2023-09-04   
855          855       855                 4          71  2023-09-18   
859          859       859                 4          70  2023-09-18   
865          865       865                 4          31  2023-09-19   
867          867       867                 4          31  2023-09-19   
870          870       870                 4          32  2023-09-19   
946          946       946                 4          71  2023-11-02   
947          947       947                 4           0  2023-11-02   
949          949       949                 4           1  2023-11-02   
950          950       950                 4          72  2023-11-04   
952          952       952                 4          58  2023-11-10   
991          991       991                 4          58  2023-12-08   
994          994       994                 4          70  2023-12-18   
1051        1051      1051                 4          56  2022-11-09   
1158        1158      1158                 4          50  2024-06-27   
1163        1163      1163                 4           1  2024-03-20   
1170        1170      1170                 4          67  2024-06-30   
1647        1647      1647                 4           6  2025-04-22   
1675        1675      1675                 4          71  2025-05-20   
1676        1676      1676                 4          22  2025-05-20   
1770        1770      1770                 4          26  2025-08-15   

     delivery_date        amount    qty  valid_delivered_qty     unit_price  
3       2020-03-29  5.144759e+06   25.0                 12.0  205790.366021  
33      2020-09-18  5.173288e+05    8.0                  7.0   64666.094581  
378     2021-11-03  1.046714e+06   20.0                 10.0   52335.710560  
432     2022-05-04  1.501567e+07   35.0                 20.0  429019.241050  
433     2022-03

**Категория 4:** В начале марта 2022 года разорвались логистические цепочки авиазапчастей. Тот факт, что задержки по этим заказам растянулись до мая, говорит о том, что поставщики метались в поисках путей обхода санкций.Это сильно испортило статистику.

**Итоги:**  
**Категория 1:** Единственная категория с нулевым уровнем недопоставок. Поставщики работают идеально.  
**Категория 3:** Несмотря на огромные объемы (тысячи единиц недовоза), в процентах это самая надежная категория после первой (99.2% успеха).Можно минимизировать складские площади, так как риск срыва поставки крайне мал.
**Категория 2:** Средний риск, но данные 2025 года показывают «дыры» в поставках. Проблема в том, что поставщики перестали справляться с объемами.Вместо одной крупной заявки, которая вызывает отказ, следует отправлять несколько мелких заказов в разное время.
**Категория 0:** Очень надежна по объему доставляемых товаров, но имеет частые срывы (11%), скорее всего это связано с распиздяйством поставщиков.
**Категория 4:** Худший показатель по доставленным товарам (95.6%). Самая высокая чувствительность к внешним кризисам (2022 год).Здесь нельзя полагаться на оперативную доставку в условиях геополитеческой нестабильности. Также надо рассчитывать объем заказа с учетом потерь в 5% т.е. заазывать чуть бульше необходимого.